src/config.ts

```diff
@@ -109,7 +121,15 @@ export function setCurrentDate(date = new Date()): void {
   currentDate = date;
 }
 
+// The config is used as a cache key for other operations; for example the pages
+// are used as a cache key for search indexing and the previous & next links in
+// the footer. When given the same spec (because import returned the same
+// module), we want to return the same Config instance.
+const configCache = new WeakMap<any, Config>();
+
 export function normalizeConfig(spec: any = {}, defaultRoot = "docs"): Config {
+  const cachedConfig = configCache.get(spec);
+  if (cachedConfig) return cachedConfig;
```

Comment:
Do you prefer this pattern over:
```
  if (configCache.has(spec)) return configCache.get(spec);
```
For performance maybe?

Reply:
For TypeScript.

Previously you had declared the `WeakMap` without types, so keys and values were type `any` by default. In this case TypeScript considers the return value of `map.get` as `any`, too, and hence doesn’t require that it returns a defined value and allows `map.get(key).property` even though it is unsafe.

When we declare the cache as `WeakMap<Config["pages"], {json: string; freshUntil: number}>`, then TypeScript will enforce type checking because `map.get(key)` now returns `{json: string; freshUntil: number} | undefined`. So if we don’t use the pattern above, we get an `Object is possibly 'undefined'` type error:

<img width="898" alt="Screenshot 2024-03-23 at 9 07 57 AM" src="https://github.com/observablehq/framework/assets/230541/32ec69ce-c2ff-423a-8c35-965c552ba40a">

TypeScript isn’t smart enough to know that if `map.has(key)` returns true, then a subsequent `map.get(key)` with the same `key` returns a defined value. (There’s no way to declare such a relationship in an interface, at least not yet.) So as a type-safe alternative, we call `map.get(key)` first, and then test the returned value which can’t change, and hence TypeScript knows that the value is defined.